### ISSUE TITLE
[UX] Rethink the menu organization/structure in Backdrop 2.x

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#2118]
I think we should re-examine the menu structure again in Backdrop 2.x.

I've been spending a lot of time in Drupal 6 (working on upgrading to Backdrop, of course) and what made that top-level menu more intuitive than Drupal 7's was the mystery menu named "Site building". Yes, that name didn't really mean anything to anyone, but all the important things you needed to **build** for your **site** lived in there: Blocks, Menus, Image styles, URL aliases, Panels, Views. Not all of these things are structural! And, some of the more structural things (like taxonomy) lived under content (which also makes more sense, imho), since terms are fieldable entities in D7, they are now content more than ever.

Why do I find D6 more intuitive than what we have in D7/BD?  Well, all the things that just had simple settings-forms were under configuration (okay, except for maybe input formats). All the interfaces under site building were far more complicated and were generally for creating and managing *\* other things*\* of that type (blocks, menus, views, panels, etc). It became clear with use which place you would go to do what.

I also really like the idea of grouping modules, themes, and layouts together under some kind of name like "Add-ons" or "Plugins", especially now that we have project browser in core. I'm guessing this is what "Site building" was for initially in D6, but contrib took over and started putting other important things in there too.

TL;DR: Proposed changes:
- move modules, themes & layouts together under "Add-Ons" (or something)
- move Taxonomy back under content
- reconsider "Structure" as a top-level category

**Comment by mikemccaffrey** [comment]
I'm pretty sure I've suggested this exact thing in another ticket. However, welcome to the party!

**Comment by jenlampton** [comment]
I know the modules, themes, layouts grouping idea was yours, but I couldn't find any other issue for it. Better too many than too few? :)

**Comment by olafgrabienski** [comment]
Actually, I prefer the menu structure in Drupal 7 (or D8, and current Backdrop which are similar). Sure, building more and more D7 sites I became familiar with it, but I remember that I liked it from the beginning more than the menu structure in D6.

So far, I'm not very convinced of putting modules, themes and layouts together under one category, and I don't want Taxonomy to be moved back under content.

Which doesn't mean that the current situation is perfect. I totally agree to re-examine the menu structure, looking forward to take part in the more detailed discussion ...

**Comment by jenlampton** [comment]
Do you have reasons why you prefer one over the other?

**Comment by olafgrabienski** [comment]
Yes, I have reasons, but maybe not aware of all of them in this moment. For a more detailed discussion we could do some preparation, e.g. a tabular comparison of the menu structure in D-6-7-8, Backdrop and even WordPress. Some first thoughts only:

"Content" in D6

- From a site builder perspective, in Drupal 6, too many items were under "Content". Not only taxonomy but also content types. In my opinion, for a site builder, both fitted better in "Structure" of Drupal 7 (or "Site building", as in D6).
- I agree, that taxonomy has to do with content and that taxonomies (or vocabularies and terms) "are" content but they serve also to structure and build the site.
- Not thinking about the differences of core and contrib modules at that time, I found it annoying that some modules in D6 like "Backup and Migrate" lived under "Content". The new place in D7 (Configuration > System, when I remember correct) seemed more logical to me. As this place existed only in D7, this contrib module could find finally a better place.

"Add-ons" or "Plugins" for modules, themes and layouts in Backdrop

- The names sound very general, or too abstract in my opinion.
- Modules and themes for example are both "Projects", but very different ones.
- Besides, in D6 modules and themes were in "Site building" (Subtitle: Control how your site looks and feels.). If we put them in "Add-ons" or so in Backdrop, we can't put anymore "all the important things you need to build for your site" in "Site building".

**Comment by jenlampton** [comment]
> in Drupal 6, too many items were under "Content"

Good point, I do think it would be important to keep the number of items to a minimum under content. The items I was envisioning there was something like: Boxes, Comments, Categories, Pages.

>  agree, that taxonomy has to do with content and that taxonomies (or vocabularies and terms) "are" content but they serve also to structure and build the site.

Let's pretend we didn't have a section labeled "Structure". What heading would you put them under then? (just for fun:)

> I found it annoying that some modules in D6 like "Backup and Migrate" lived under "Content". The new place in D7 (Configuration > System, when I remember correct) seemed more logical to me.

I also prefer all the nesting we have under Configuration. Giving contrib more options for where to put their things is good, IMHO.

> Modules and themes for example are both "Projects", but very different ones.

So are layouts, install profiles, and distributions. But rather than thinking about what these things are actually called, let's think about what term a user may be scanning for when they want to add something to their site that is not there yet. Something like "App store" could even work for this.

> If we put them in "Add-ons" or so in Backdrop, we can't put anymore "all the important things you need to build for your site" in "Site building".

Why not?  "Installing things" and "building things" are different experiences for the user. One is choosing and copying code, the other is clicking together configurations. They are two very different operations and should be treated (and organized?) differently. I don't know what to call either of those operations, but I think it would be a better experience if we treated (and organized?) them separately rather than lumped them together.

**Comment by klonos** [comment]
> Something like "App store" could even work for this.

Dangerous ground there. Big non-no from me 👎 I'd prefer something like "Extend" or "Add features" instead.

> - move modules, themes & layouts together under "Add-Ons" (or something)

I've filed #1701 back in March as a Project Browser follow-up. I then separated that into two related sub-tasks:

- #1780: merge module/them/layout _manage_ pages under a single top-level menu
- #1781: merge module/them/layout _update_ pages under a single "Updates" page, under the same top-level menu item as #1780

**Comment by klonos** [comment]
With the proposed changes in #1701, #1780 and #1781, we'd end up with something like this:

- Extend Backdrop (exact name TBD) <-- single **top-level menu item** to hold _ALL_ project-related tasks (install/manage/update)
  - Install projects <-- single **primary tab** to hold Project Installer for _ALL_ projects (modules/themes/layouts)
    - Install modules <-- **"pseudo" item**: filters the "Install projects" page to show _only_ modules
    - Install themes <-- **"pseudo" item**: filters the "Install projects" page to show _only_ themes
    - Install layouts <-- **"pseudo" item**: filters the "Install projects" page to show _only_ layouts
  - Manage modules <-- single **primary tab** for enabling/disabling installed modules
  - Manage themes <-- single **primary tab** for enabling/disabling installed themes and setting the defaults for front-end and admin backend
  - Manage layouts <-- single **primary tab** that holds the Layout UI
  - Updates <-- single **primary tab** that lists available updates for core and all project types (modules/themes/layouts)

Paths:

`/extend`
  `/extend/install`
    `/extend/install/modules`
    `/extend/install/themes`
    `/extend/install/layouts`
  `/extend/manage/modules`
  `/extend/manage/themes`
  `/extend/manage/layouts`
  `/extend/update`

**Comment by jenlampton** [comment]
> Manage layouts <-- single primary tab that holds the Layout UI

I adamantly disagree with this.  "building" anything in the same place where things are "installed" is a huge no-no. Layout UI needs to stay under `Site building` / `Structure` / whatever or we are going to have an even worse nightmare than we do now with people being confused about the three things a `Layout` can be.

If we want to put the layout settings page here (the place where you choose which layouts are in use for a site) that would be okay, and is inline with modules & themes. But no way are we moving Layout UI into the installer section.

**Comment by klonos** [comment]
> I adamantly disagree with this.  "building" anything in the same place where things are "installed" is a huge no-no. It's a clash of concepts as well as a clash of actions. Layout UI needs to stay under `Site building` / `Structure` / `Whatever` or we are going to have an even worse nightmare than we do now with people being confused about the three things a `Layout` can be.

Yeah, I kinda agree with what you said @jenlampton. At the very least, you have some very valid points there. Unless I come up with something "better" that addresses your concerns that is.

**Comment by jenlampton** [comment]
> Unless I come up with something "better"

We have time! :)

**Comment by olafgrabienski** [comment]
Reading https://github.com/backdrop/backdrop-issues/issues/1701#issuecomment-271031564 of @Dinornis I remembered this discussion about a possible future menu reorganization. Although we have time to discuss it later, I want to share another approach regarding the location of *Themes*, *Modules* and *Layouts* in the admin menu.

In my opinion it's not necessary to hide these important items under abstract terms. Currently we have:

- Home
- Content
- User accounts
- **Appearance** (= Themes)
- **Functionality** (= Modules)
- Structure > **Layouts** > ... > Manage blocks
- Configuration
- Reports

While *Structure* holds many entries besides *Layouts*, the terms *Appearance* and *Functionality* each work just as an abstract wrapper for the concrete thing. Are the abstract wrapper terms maybe meant to hold additional things in the future (like in *Functionality > Modules* vs. *Functionality > Plugins*)? And if yes, is it likely that there will be a real need for that?

I'm asking because I'd like to move the hidden yet important items up to the first level. I guess that *Themes* and *Modules* are quite popular and understandable terms, also to newcomers (coming for instance from Drupal or WordPress).

In comparison, the concept of *Layouts* may not be so familiar for many users so far. As Layouts are very important in Backdrop, they would profit a lot and conceptually rightly of being a top level item. (As a side effect, the *Manage blocks* items would also move up one level which might be a helpful as well.)

So, here's my suggestion:

- Home
- Content
- User accounts
- **Modules**
- **Themes**
- **Layouts**
- Structure
- Configuration
- Reports

**Comment by Dinornis** [comment]
👍 I like this proposal. Only maybe in a different order.
> * Home
> * Content
> * Structure (not sure about this term, but can't think of better one)
> * **Layouts** (Blocks could possibly live in here)
> * **Modules**
> * **Themes**
> * User accounts
> * Configuration
> * Reports

**Comment by olafgrabienski** [comment]
Thanks! Yes, I didn't care much about the order so far. The only idea was that *Themes* and *Layouts* would be good neighbors because both handle the visual presentation of a site. *Layouts* and *Structure* can also be nearby as *Layouts* currently are even a part of the *Structure* sub menu.

What are the reasons for your specific order?

**Comment by Dinornis** [comment]
>What are the reasons for your specific order?

Most frequently used first.  You are right though, _Themes_ and _Layouts_ would be good neighbors because both handle the visual presentation of a site.

Also, should the _Administration bar_ page at _admin/config/administration/admin-bar_ include a sentence about the fact that the user is able to customise the _Administration_ menu at _/admin/structure/menu/manage/management_ ?

**Comment by jlfranklin** [comment]
> If we want to put the layout Settings page here (the place where you choose which layouts are in use for a site) that would be okay, and is inline with what happens with modules & themes. But no way are we moving the whole Layout UI into the installer section.

The 👍 reaction above is for the whole of @jenlampton 's comment above, and in particular for this quoted bit.  I like the idea of putting all the menu items about adding and enabling extensions, be they modules, themes, layouts, libraries, etc, in one place.  Unless I misunderstand something about them, layouts are all about the presentation layer, not the data layer, so I would like to see the Layout UI move out from Structure to a menu more focused on the look and feel of a site.

**Comment by klonos** [comment]
Here's an idea:

- Introduce a new `Extend Backdrop` menu. This menu will hold a single Project Browser instance for all project types but offer a filter so that people can browse only themes, modules
- Move `Appearance` -> `List themes` under the new `Extend Backdrop` menu, and rename to "Manage themes" instead
- Move `Functionality` -> `List modules` under the new `Extend Backdrop` menu, and rename to "Manage modules"
- Keep `Structure` -> `Layouts` -> `List layouts` where it is, but rename to "Manage layouts"
- Move `Structure` -> `Layouts` -> `Settings` under the new `Extend Backdrop` menu and rename to "Manage layout templates"
- Merge all `Update [project type]` and `Reports` -> `Available updates` pages into the Project Browser (see #2714)

- `Extend Backdrop`
-- `Install and Update Projects` (perhaps since is a talk about 2.x, also take the leap and rename to "Plugins" or "Add-ons")
-- `Manage themes`
-- `Manage modules`
-- `Manage layout templates`

**Comment by jenlampton** [comment]
> layouts are all about the presentation layer, not the data layer

This is why I dislike the word "structure" so much, everyone has their own definition of what it means. Views don't alter the data layer of a website, but fields do, and fields are under content. The structure section of the website is for site building or architectural tasks - in Drupal 6 it was accurately named "Site building". Layouts and views are both definitely site-builder tools.

**Comment by klonos** [comment]
>>  Introduce a new Extend Backdrop menu

>I could get behind this idea...

It's already started to happen: https://backdropcms.org/user-guide/add-ons-modules-themes-and-layouts 😄 

>> This menu will hold a single Project Browser instance for all project types

> ...but not this one. There is no use case for looking across all projects for something called "flag", for example, and if we allowed that it would make for a very confusing user experience.

The use case I had in mind was not to allow searching by using the same term across all project types. The goal here is to save users clicks and having them navigate from section to section only to perform a similar task from a similar UI. Consider this scenario:

User wants to install a module, a theme and a layout (or multiple of them, but lets consider one of each project type for simplicity's sake). Steps they have to take now:

1. Navigate to `admin/modules/install`
2. Type name of module and search
3. Click add to queue
4. Navigate to `admin/appearance/install` (which by the way should be `admin/themes/install` but won't matter once we merge these pages 😛 )
5. Type name of theme and search
6. Click add to queue
7. Navigate to `admin/structure/layouts/install`
8. Type name of layout and search
9. Click add to queue
10. Click install

Same process if the Project Browser was a single thing:

1. Navigate to `admin/extend`
2. Type name of module and search
3. Click add to queue
4. Type name of theme and search
5. Click add to queue
6. Type name of layout and search
7. Click add to queue
8. Click install

Now, 10 steps vs. 8 steps might not seem much, but keep in mind that "Navigate to ..." involves multiple hover-overs in the admin bar + page refresh = more time required to accomplish the same task = UX-. Also people might be out of their mind doing admin tasks on their phones, but it won't be too uncommon to be using their tablets. On those devices "Navigate to ..." involves multiple clicks and when the device is in portrait mode where the admin bar switches to mobile mode, additional clicks = UX--

Lastly, consider the "Oh, I forgot I need module xyz" scenario where the user needs to return back to the module installation page from the theme or layout installation page. More clicks, more page reloads and more pogo-sticking around the admin UI.

For those that need to separate themes from modules and from layouts in that list, we'll have a series of checkboxes which they can use to filter it by.

> ... If you install both a module and a theme called "flag" your site might explode. :(

That literally cannot happen, since we cannot possibly have two projects that have the same name in backdrop-contrib. If the user manually installs a custom project with the same name as an official contrib project, then bad things will happen whether we have the Project Browser be a single, unified thing or 3 separate things split among various sections throughout the admin UI.

**Comment by jenlampton** [comment]
Okay, so it's not that you want a single browser that searches across all projects, but to have the task performed on the same page. That makes sense. The first Question could be "which thing are you searching for?" and then we could search only that kind of project.

**Comment by jenlampton** [comment]
Rereading some of the comments from above.. 

> While Structure holds many entries besides Layouts, the terms Appearance and Functionality each work just as an abstract wrapper for the concrete thing. Are the abstract wrapper terms maybe meant to hold additional things in the future?

No, there was no intent for these items to hold anything other than what is there now.

These two changes came directly from usability studies done on both Drupal 6 and Drupal 7. The words "Modules" and "Themes" make sense only to experienced Drupal (or CMS) users and not to the majority of people. People want to change the "appearance" of their site. 

The link for "Appearance" was moved out of "site building" and into the top-level in D7 because it's usually the first thing a new person wants to change about their site. Giving people an easy win early makes them feel better about the software.

People also have no idea what a "module" is, and when people don't know the meaning of a word, using it in navigation fails. That particular link in the menu was invisible to any participant who didn't have previous experience with Drupal, Joomla, or Unix, their eyes moved right over it as though it wasn't there.

The link for "Modules" was moved to the top-level in D7, and had its name changed in D8/Backdrop because it's arguably the most important and powerful feature.  All were attempts to make the modules page more discoverable to first-time users.

I personally like the idea of adding a new top-level item called "Add-ons" and putting the project browser/updater in there, but I'm torn about moving the management of those things. Organizationally, it makes sense. But from a UX perspective, I'm not sure they belong in the same place. 

I'd like to do some usability studies of before/after to make sure we aren't causing any UX regressions by burying two of the most important tasks where people can't find them at first glance, and back behind terms people don't understand, again.

**Comment by olafgrabienski** [comment]
> These two changes came directly from usability studies done on both Drupal 6 and Drupal 7.

Interesting. Do you know if the usability studies are available somewhere?

My idea to put *Themes* and *Modules* to the top-level without changing their names to abstract terms like *Appearance* and *Functionality* was partially inspired by WordPress: I remembered vaguely that in WordPress *Plugins* was a top-level item, i.e. without wrapping it with an abstract term. Today I had a look at a site hosted on wordpress.com to check if I remembered it correctly, and that was even more interesting:

In the (old) "WP Admin" menu, there is indeed just *Plugins* on the one hand. On the other hand there is *Appearance > Themes*, like in Backdrop. In contrast to Backdrop there are however more items under *Appearance*, for example *Widgets*, *Menus*, or *Header*. Here is a screenshot:

![screen-wordpress-menu-wp-admin](https://user-images.githubusercontent.com/14188307/26849640-4dfd6f5c-4b05-11e7-94a8-b0e0bbf8fdd7.png)

The WP Admin menu isn't the default on wordpress.com however. In the *new* admin menu, you see *Plugins* and *Themes* as top-level items, so they put *Themes* out of *Appearance*, see a screenshot:

![screen-wordpress-menu](https://user-images.githubusercontent.com/14188307/26850115-bbfe63ac-4b06-11e7-9e82-a9ddc790aabc.png)

I don't know the background of the WordPress changes. In my opinion *Appearance* and *Functionality* are however both very general, maybe too general. One example: A person who wants to change the 'appearance' of the main menu won't find a respective setting under *Appearance > Themes* (but has to change the respective block in one ore more *Layouts*).

That said, I agree that doing usability studies would be great. I'd like to discuss then also which 'personas' make sense for Backdrop.

**Comment by BWPanda** [comment]
Regarding the Content menu, it currently has links to both add new, and manage existing, Nodes (Content) and Files. I think we can add Add and Manage links for Custom Blocks and Taxonomy Terms as well.

And just as the link for managing content types lives elsewhere, so can the link for managing vocabularies (and custom blocks don't really have different 'types' to worry about).

**Comment by BWPanda** [comment]
Duplicate of #1149 (or the other way around)?

**Comment by BWPanda** [comment]
Just marked (and closed) https://github.com/backdrop/backdrop-issues/issues/1149 as a duplicate of this issue.

**Comment by klonos** [comment]
Tweaking the issue title so it's easier to find + also switching to the 2.x milestone, since that's what's in the issue title (feel free to switch to 1.x, but then also please rename the issue title).
